mesonbuild/interpreter/interpreter.py

```diff
@@ -3135,29 +3135,31 @@ def do_validate_within_subproject(norm):
                 inputtype = 'directory'
             else:
                 inputtype = 'file'
-            if InterpreterRuleRelaxation.ALLOW_BUILD_DIR_FILE_REFERENCES in self.relaxations and builddir in norm.parents:
+            if InterpreterRuleRelaxation.ALLOW_BUILD_DIR_FILE_REFERENCES in self.relaxations and path_starts_with(norm, self.environment.build_dir):
                 return
-            if srcdir not in norm.parents:
+
+            if not path_starts_with(norm, self.environment.source_dir):
                 # Grabbing files outside the source tree is ok.
                 # This is for vendor stuff like:
                 #
                 # /opt/vendorsdk/src/file_with_license_restrictions.c
                 return
-            project_root = Path(srcdir, self.root_subdir)
-            subproject_dir = project_root / self.subproject_dir
-            if norm == project_root:
-                return
-            if project_root not in norm.parents:
-                raise InterpreterException(f'Sandbox violation: Tried to grab {inputtype} {norm.name} outside current (sub)project.')
-            if subproject_dir == norm or subproject_dir in norm.parents:
-                raise InterpreterException(f'Sandbox violation: Tried to grab {inputtype} {norm.name} from a nested subproject.')
+
+            project_root = os.path.join(self.environment.source_dir, self.root_subdir, '')
```

Comment:
Why do we need the `''` at the end? The only difference is a trailing slash in the returned path string but path_starts_with shouldn't care about that...

Reply:
Yes it's a relic of when I tried just using startswith. Will fix.